src/core/vectortile/qgsvectortilelayerrenderer.cpp

```diff
@@ -132,7 +132,7 @@ bool QgsVectorTileLayerRenderer::render()
   // add @zoom_level variable which can be used in styling
   QgsExpressionContextScope *scope = new QgsExpressionContextScope( QObject::tr( "Tiles" ) ); // will be deleted by popper
   scope->setVariable( QStringLiteral( "zoom_level" ), mTileZoom, true );
-  scope->setVariable( QStringLiteral( "vector_tile_zoom" ), mTileMatrixSet.scaleToZoom( tileRenderScale ), true );
+  scope->setVariable( QStringLiteral( "vector_tile_zoom" ), QgsVectorTileUtils::scaleToZoom( tileRenderScale ), true );
```

Comment:
This does need to be determined by the matrix set. There's different logic applied depending on the tile source (ie ESRI sources use slightly different scale to zoom calculations)

Reply:
What about using zoom levels up to 30 in the `scaleToZoom` instead of checking only tile matrices present in the layer's tile matrix set?